pkg/jobworker/worker.go

```diff
@@ -0,0 +1,219 @@
+
+// NewJob initialises a Job
+func NewJob(id string, cmd *exec.Cmd, con ResourceController) *Job {
+	return &Job{ID: id, running: true, cmd: cmd, con: con, done: make(chan bool, 1)}
+}
+
+// Start calls start using the default ResourceController Cgroup
+func Start(opts JobOpts, cmd string, args ...string) (j *Job, err error) {
+	return StartWithController(&Cgroup{"/sys/fs/cgroup"}, opts, cmd, args...)
+}
+
+// StartWithController creates a job's cgroup, adds the resource controls from opts, creates a log file for the cgroup and
+// set's it to the exec.Cmd STDOUT and STDERR. Finally we Start the exec.Cmd and start a go routine to handle the blocking
+// call to Wait so that we can update the job's running flag.
+func StartWithController(con ResourceController, opts JobOpts, cmd string, args ...string) (j *Job, err error) {
+	// Create the job
+	j = NewJob(uuid.New().String(), exec.Command(cmd, args...), con)
+	// Create the cgroup and configure the controllers
+	if err = j.con.CreateGroup(j.ID); err != nil {
+		return nil, fmt.Errorf("failed to create cgroup: %w", err)
+	}
+	// Update cgroup controllers to add resource control to process
+	if err = j.con.AddResourceControl(j.ID, opts); err != nil {
+		return nil, fmt.Errorf("failed to add resource control: %w", err)
+	}
+	// Add job's process to cgroup
+	if err = j.con.AddProcess(j.ID, j.cmd); err != nil {
+		return nil, fmt.Errorf("failed to add PID to cgroup: %w", err)
+	}
+	defer syscall.Close(j.cmd.SysProcAttr.CgroupFD)
+	// Don't inherit environment from parent
+	j.cmd.Env = []string{}
+	// Pipe STDOUT and STDERR to a log file
+	f, err := os.OpenFile(logPath(j.ID), os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open job's log file: %w", err)
+	}
+	j.cmd.Stdout = f
+	j.cmd.Stderr = f
+	// Run the command as a given user as not to escalate privilege, since the executing user must also manage cgroups
+	if WORKER_UID != -1 && WORKER_GUID != -1 {
+		j.cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(WORKER_UID), Gid: uint32(WORKER_GUID)}
+	}
+	j.cmd.SysProcAttr.Setpgid = true
+	// Start the job
+	err = j.cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start job's exec.Cmd: %w", err)
+	}
+	// Assign the process group ID to the job so that we have a reference to signal child processes in Stop if the command quits
+	j.pgid, err = syscall.Getpgid(j.cmd.Process.Pid)
+	if err != nil {
+		fmt.Printf("error getting pgid: %v", err)
+		return nil, err
+	}
+	// Run go routine to handle the blocking call exec.Cmd.Wait() and update the running flag to indicate the job has complete
+	go func(runningJob *Job, logFile *os.File) {
+		runningJob.cmd.Wait()
+		runningJob.setRunning(false)
+		runningJob.done <- true
+		logFile.Close()
+	}(j, f)
+	return j, nil
+}
+
+// Stop request a job's termination using SIGTERM and deletes it's cgroup. If the SIGTERM is ignored, we send
+// a SIGKILL after STOP_GRACE_PERIOD.
+func (job *Job) Stop() error {
```

Comment:
Given that this can potentially take up to `STOP_GRACE_PERIOD`, is it worth taking a `Context` argument so that the caller can abort if they want?

Reply:
Yes that's a good point